# 2장 1강: t-검정의 이론과 가정 — 실습문제

## 실습 목표

- 평균 차이와 표준오차를 이용해 t통계량을 직접 계산할 수 있다.
- 단일표본 t검정으로 하나의 표본평균과 기준값을 비교할 수 있다.
- 독립성·정규성·등분산성을 확인한 뒤 독립표본 t검정을 수행할 수 있다.
- 독립표본과 대응표본 상황을 구분하고 적절한 함수를 선택할 수 있다.
- p-value를 유의수준과 비교하여 검정 결과를 올바르게 해석할 수 있다.

## 실습 환경 / 데이터

- Python
- NumPy, pandas
- scipy.stats
- `ames_housing.csv`

| 컬럼 | 의미 |
|---|---|
| `SalePrice` | 주택 판매가격 |
| `KitchenQual` | 주방 품질 |

> 모든 검정은 양측검정이며 유의수준 `α = 0.05`를 사용합니다.  
> Ames 표본 추출에는 지정된 `random_state`를 사용해 결과를 재현합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 전체 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [1]:
# 실습 준비 코드를 작성하세요.

import numpy as np
import pandas as pd
from scipy import stats

df = pd.read_csv('ames_housing.csv', encoding='utf-8-sig')

print('데이터 크기 :', df.shape)
print('전체 결측치 수 :', df.isnull().sum().sum())
df.head()

데이터 크기 : (1460, 10)
전체 결측치 수 : 0


,SalePrice,GrLivArea,LotArea,OverallQual,KitchenQual,CentralAir,HeatingQC,PavedDrive,Neighborhood,YearBuilt
0,208500,1710,8450,7,Gd,Y,Ex,Y,CollgCr,2003
1,181500,1262,9600,6,TA,Y,Ex,Y,Veenker,1976
2,223500,1786,11250,7,Gd,Y,Ex,Y,CollgCr,2001
3,140000,1717,9550,7,Gd,Y,Gd,Y,Crawfor,1915
4,250000,2198,14260,8,Gd,Y,Ex,Y,NoRidge,2000


---

## 필수 1. t통계량 직접 계산과 단일표본 t검정

### 문제 1-1. 평균 판매가격은 180,000달러와 다른가?

#### 문제 설명

Ames 주택 30개를 표본으로 추출하여 평균 판매가격이 기준값 180,000달러와 다른지 확인합니다. 먼저 t통계량을 직접 계산한 뒤 `ttest_1samp()` 결과와 비교하세요.

#### 요구사항

1. `SalePrice`에서 `n=30`, `random_state=2`로 표본을 추출하여 `sale_sample`에 저장하세요.
2. 표본 수, 표본평균, 표본표준편차를 출력하세요.
3. `stats.sem()`을 이용해 표준오차를 계산하세요.
4. 다음 식으로 t통계량을 직접 계산하세요.  
   `t = (표본평균 - 기준값) / 표준오차`
5. Shapiro-Wilk 검정으로 표본의 정규성을 확인하세요.
6. 다음 가설을 작성하세요.
   - H₀: 모집단 평균 판매가격은 180,000달러이다.
   - H₁: 모집단 평균 판매가격은 180,000달러가 아니다.
7. `stats.ttest_1samp()`로 단일표본 t검정을 수행하세요.
8. 직접 계산한 t통계량과 함수가 반환한 t통계량을 비교하세요.
9. p-value를 이용해 귀무가설 기각 여부를 판단하세요.

#### 해석 질문

**Q1.** t통계량은 평균 차이와 표준오차를 어떻게 이용한 값인가요?  
**Q2.** 같은 평균 차이라면 표준오차가 작아질수록 t통계량의 절댓값은 어떻게 변하나요?  
**Q3.** 직접 계산한 t통계량과 `ttest_1samp()`의 t통계량은 일치하나요?  
**Q4.** 검정 결과 평균 판매가격이 180,000달러와 다르다고 판단할 수 있나요?

#### 제출 결과

- 기술통계량과 정규성 결과
- 직접 계산한 표준오차와 t통계량
- 단일표본 t검정 결과
- 귀무가설 판단과 해석
- Q1~Q4 답변

In [2]:
# 필수 1 코드를 작성하세요.

sale_sample = df['SalePrice'].sample(n=30, random_state=2)

print('[기술통계]')
print(f'표본 수 : {len(sale_sample)}')
print(f'표본 평균 : {sale_sample.mean().round(3)}')
print(f'표본 표준편차 : {stats.sem(sale_sample).round(3)}')


# t통계량 직접 계산
mu0 = 180000    # 기준값

t_stat_ind = (sale_sample.mean() - mu0) / stats.sem(sale_sample)

print('\n[t통계량]')
print(f't통계량 : {t_stat_ind:.3f}')


# Shapiro-Wilk 정규성 검정
shapiro_stat, shapiro_pvalue = stats.shapiro(sale_sample)

print('\n[Shapiro-Wilk 정규성 검정]')
print(f'검정통계량 : {shapiro_stat:.3f}')
print(f'p-value : {shapiro_pvalue:.3f}')
print('p-value가 0.05 이상이므로 정규성 가정을 만족합니다.' if shapiro_pvalue >= 0.05 else 'p-value가 0.05 미만이므로 정규성 가정을 만족하지 않습니다.')

# 가설 정립
# - H₀: 모집단 평균 판매가격은 180,000달러이다.
# - H₁: 모집단 평균 판매가격은 180,000달러가 아니다.

# 독립표본 t검정
print('\n[독립표본 t검정]')

t_stat, pvalue = stats.ttest_1samp(sale_sample, popmean=mu0)
print(f't검정 : {t_stat:.3f} | p-value : {pvalue:.3f}')
print('p-value가 0.05 미만이므로 귀무가설을 기각 가능합니다.' if pvalue <= 0.05 else 'p-value가 0.05 이상이므로 귀무가설이 기각 불가능합니다.')

# 직접 계산 vs ttest_1smp
print(f'\n직접 계산 t통계량 : {t_stat_ind:.3f} | ttest_1samp t통계량 : {t_stat:.3f}')

[기술통계]
표본 수 : 30
표본 평균 : 233323.8
표본 표준편차 : 15703.756

[t통계량]
t통계량 : 3.396

[Shapiro-Wilk 정규성 검정]
검정통계량 : 0.954
p-value : 0.213
p-value가 0.05 이상이므로 정규성 가정을 만족합니다.

[독립표본 t검정]
t검정 : 3.396 | p-value : 0.002
p-value가 0.05 미만이므로 귀무가설을 기각 가능합니다.

직접 계산 t통계량 : 3.396 | ttest_1samp t통계량 : 3.396


### 필수 1 답변 작성란

- **Q1.** t통계량은 평균 차이와 표준오차를 어떻게 이용한 값인가요?  
    - 관측된 평균과 기준값의 차이를 (그 차이의 불확실성을 의미하는) 표준오차로 나눈 값

- **Q2.** 같은 평균 차이라면 표준오차가 작아질수록 t통계량의 절댓값은 어떻게 변하나요?  
    - 분모(표준오차)가 작아지므로 T통계량의 절댓값은 커진다.

- **Q3.** 직접 계산한 t통계량과 `ttest_1samp()`의 t통계량은 일치하나요?
    - 일치한다. 두 방법 모두 약 3.396으로 계산된다.

- **Q4.** 검정 결과 평균 판매가격이 180,000달러와 다르다고 판단할 수 있나요?
    - 그렇다. p-value가 약 0.002로 0.05보다 작아 귀무가설을 기각한다.
    - 따라서 평균 판매가격은 180,000달러와 다르다.

---

## 필수 2. 독립표본 t검정의 가정 점검과 수행

### 문제 2-1. 주방 품질 `Gd`와 `TA` 집단의 평균 판매가격 비교

#### 문제 설명

주방 품질이 `Gd`인 주택과 `TA`인 주택에서 각각 20개를 추출해 평균 판매가격을 비교합니다. 두 집단은 서로 다른 주택으로 구성되어 있습니다.

#### 요구사항

1. `KitchenQual == "Gd"`와 `KitchenQual == "TA"` 집단의 `SalePrice`에서 각각 `n=20`, `random_state=42`로 표본을 추출하세요.
2. 두 집단의 표본 수와 평균을 확인하세요.
3. 데이터 수집 구조를 근거로 두 집단의 독립성을 설명하세요.
4. 각 집단에 Shapiro-Wilk 정규성 검정을 수행하세요.
5. 두 집단에 Levene 등분산 검정을 수행하세요.
6. 가정 점검 결과에 따라 `equal_var=True` 또는 `equal_var=False`를 결정하세요.
7. `stats.ttest_ind()`로 독립표본 t검정을 수행하세요.
8. t통계량과 p-value를 출력하고 두 집단 평균 차이를 해석하세요.

#### 해석 질문

**Q1.** 두 집단이 독립표본인 이유는 무엇인가요?  
**Q2.** 독립성은 별도의 p-value로 확인할 수 있나요?  
**Q3.** 정규성과 등분산성 가정은 각각 충족되나요?  
**Q4.** `equal_var`에는 어떤 값을 사용해야 하나요?  
**Q5.** 두 집단의 평균 판매가격에는 통계적으로 유의한 차이가 있나요?

#### 제출 결과

- 집단별 표본 수와 평균
- 독립성 설명
- 정규성 및 등분산성 검정 결과
- `equal_var` 선택 근거
- 독립표본 t검정 결과와 해석
- Q1~Q5 답변

In [3]:
# 필수 2 코드를 작성하세요.

import numpy as np

group_gd = df[df['KitchenQual'] == 'Gd']['SalePrice'].sample(n=20, random_state=42)
group_ta = df[df['KitchenQual'] == 'TA']['SalePrice'].sample(n=20, random_state=42)

print('[기술통계]')
print(f'표본 수 - Gd  : {len(group_gd)} | TA : {len(group_ta)}')
print(f'표본 평균 - TA  : {group_ta.mean().round(3)} | TA : {group_ta.mean().round(3)}')

# 독립성 검증
print('\n[독립성 검증]')
print('두 집단은 서로 다른 주방품질로 구성되어 있으므로 독립집단이다')

# Shapiro-Wilk 정규성 검정
gd_shapiro_stat, gd_shapiro_pvalue = stats.shapiro(group_gd)
ta_shapiro_stat, ta_shapiro_pvalue = stats.shapiro(group_ta)

print('\n[Shapiro-Wilk 정규성 검정]')
print(f'Gd - 검정통계량 : {gd_shapiro_stat:.3f} | p-value : {gd_shapiro_pvalue:.3f}')
print(f'TA - 검정통계량 : {ta_shapiro_stat:.3f} | p-value : {ta_shapiro_pvalue:.3f}')
print('Gd 그룹의 p-value가 0.05이상이므로 정규성을 만족합니다.' if gd_shapiro_pvalue >= 0.05 else 'Gd 그룹의 p-value가 0.05미만이므로 정규성을 만족하지 않습니다.')
print('TA 그룹의 p-value가 0.05이상이므로 정규성을 만족합니다.' if gd_shapiro_pvalue >= 0.05 else 'TA 그룹의 p-value가 0.05미만이므로 정규성을 만족하지 않습니다.')


# levene 등분산성 검증
levene_tstats, levene_pvalue = stats.levene(group_gd, group_ta)

print('\n[Levene 등분산성]')
print(f'검정통계량 : {levene_tstats:.3f} | p-value : {levene_pvalue:.3f}')
print('p-value가 0.05이상이므로 등분산성을 만족합니다.' if levene_pvalue >= 0.05 else 'p-value가 0.05미만이므로 등분산성을 만족하지 않습니다.')

# 독립표본 t검증
t_stat, pvalue = stats.ttest_ind(group_gd, group_ta, equal_var=False)

print('\n[독립표본 t검증(비등분산성)]')
print(f't통계량 : {t_stat:.3f} | p-value : {pvalue:.3f}')
print(f'p-value가 0.05 미만이므로 유의미한 차이가 존재합니다(귀무가설 기각 가능).' if pvalue < 0.05 else 'p-value가 0.05 이상이므로 유의미한 차이가 존재하지 않습니다(귀무가설 기각 불가능).')

[기술통계]
표본 수 - Gd  : 20 | TA : 20
표본 평균 - TA  : 137432.5 | TA : 137432.5

[독립성 검증]
두 집단은 서로 다른 주방품질로 구성되어 있으므로 독립집단이다

[Shapiro-Wilk 정규성 검정]
Gd - 검정통계량 : 0.955 | p-value : 0.458
TA - 검정통계량 : 0.941 | p-value : 0.250
Gd 그룹의 p-value가 0.05이상이므로 정규성을 만족합니다.
TA 그룹의 p-value가 0.05이상이므로 정규성을 만족합니다.

[Levene 등분산성]
검정통계량 : 3.781 | p-value : 0.059
p-value가 0.05이상이므로 등분산성을 만족합니다.

[독립표본 t검증(비등분산성)]
t통계량 : 3.425 | p-value : 0.002
p-value가 0.05 미만이므로 유의미한 차이가 존재합니다(귀무가설 기각 가능).


### 필수 2 답변 작성란

- **Q1.** 두 집단이 독립표본인 이유는 무엇인가요?  
    - Gd/TA 각 집단이 서로 다른 주택으로 구성되어 하나의 관측값이 다른 관측값과 짝을 이루지 않기 때문

- **Q2.** 독립성은 별도의 p-value로 확인할 수 있나요?  
    - 일반적으로 독립성은 별도의 검정  p-value로 확인하는 것이 아니라
    - 데이터의 수집 구조, Feature 엔지니어링 구조를 통해 판단한다.

- **Q3.** 정규성과 등분산성 가정은 각각 충족되나요?  
    - 필수 2 문제에서는 정규성과 등분산성 위반 근거가 없다.

- **Q4.** `equal_var`에는 어떤 값을 사용해야 하나요?  
    - levene p-value가 0.05보다 크므로 True를 사용한다.

- **Q5.** 두 집단의 평균 판매가격에는 통계적으로 유의한 차이가 있나요?
    - 유의한 차이가 있다. t 검정 결과 p-value는 약 0.002로 0.05보다 작다.
    - 일반적으로 주방 설계를 잘해놓으면 주택가격이 달라질 가능성이 있다.

---

## 과제. 대응표본과 독립표본 상황 구분

### 문제 3-1. 같은 주택의 보수 전후 예상 판매가격 비교

#### 문제 설명

부동산 회사가 동일한 주택 10채에 대해 보수 전 예상 판매가격과 보수 후 예상 판매가격을 각각 산정했습니다. 같은 위치의 값은 동일한 주택의 전후 가격으로 서로 짝을 이룹니다.

```python
before_price = [145, 162, 178, 155, 190, 172, 168, 181, 159, 175]
after_price  = [154, 170, 185, 164, 201, 179, 176, 190, 168, 183]
```

단위는 천 달러입니다.

#### 요구사항

1. 두 배열의 길이가 같은지 확인하세요.
2. 보수 전후 평균을 계산하세요.
3. 독립표본 t검정과 대응표본 t검정 중 적절한 방법을 선택하고 이유를 설명하세요.
4. 대응표본 t검정의 정규성 가정은 개별 배열이 아니라 `after_price - before_price` 차이값에 적용된다는 점을 확인하세요.
5. 차이값에 Shapiro-Wilk 정규성 검정을 수행하세요.
6. 적절한 t검정을 실행하고 t통계량과 p-value를 출력하세요.
7. 보수 전후 평균 예상 판매가격에 유의한 차이가 있는지 결론을 작성하세요.

#### 해석 질문

**Q1.** 이 데이터는 독립표본인가요, 대응표본인가요?  
**Q2.** 대응표본 t검정에서는 무엇의 정규성을 확인해야 하나요?  
**Q3.** `stats.ttest_ind()`가 아니라 어떤 함수를 사용해야 하나요?  
**Q4.** 검정 결과 보수 전후 예상 판매가격에 유의한 차이가 있나요?

#### 제출 결과

- 표본 관계 판단과 근거
- 전후 평균과 차이값
- 차이값의 정규성 결과
- 대응표본 t검정 결과
- 결과 해석
- Q1~Q4 답변

In [4]:
# 과제 코드를 작성하세요.

# 1. 두 배열의 길이가 같은지 확인하세요.
# 2. 보수 전후 평균을 계산하세요.
# 3. 독립표본 t검정과 대응표본 t검정 중 적절한 방법을 선택하고 이유를 설명하세요.
# 4. 대응표본 t검정의 정규성 가정은 개별 배열이 아니라 `after_price - before_price` 차이값에 적용된다는 점을 확인하세요.
# 5. 차이값에 Shapiro-Wilk 정규성 검정을 수행하세요.
# 6. 적절한 t검정을 실행하고 t통계량과 p-value를 출력하세요.
# 7. 보수 전후 평균 예상 판매가격에 유의한 차이가 있는지 결론을 작성하세요.


before_price = [145, 162, 178, 155, 190, 172, 168, 181, 159, 175]
after_price  = [154, 170, 185, 164, 201, 179, 176, 190, 168, 183]

# 1. 배열 길이 동일 여부
len_desc = len(before_price) == len(after_price)
print('두 배열의 길이 동일 여부 :', len_desc)

# 2. 보수 전후 평균
before_price_mean = np.mean(before_price)
after_price_mean = np.mean(after_price)

print('\n[평균]')
print('보수 전 평균 :', before_price_mean)
print('보수 후 평균 :', after_price_mean)

print('\n[검정 선택 방법]')
print('대응표본 t검정 진행')


# 3. 차이값 산정, 정규성 검증
diff = np.array(after_price) - np.array(before_price)

diff_tstats, diff_pvalue = stats.shapiro(diff)

print('\n[차이값 정규성 검증]')
print('차이값 :', diff)
print(f'검정통계량 : {diff_tstats:.4f} | p-value : {diff_pvalue:.4f}')
print('판단 : 정규성 가정을 만족합니다.' if diff_pvalue >= 0.05 else '판단 : 정규성 가정을 만족하지 않습니다.')

tstat, pvalue = stats.ttest_rel(after_price, before_price)

print('\n[대응표본 t검정]')
print(f't통계량 : {tstat:.4f}')
print(f'p-value : {pvalue:.4f}')

if pvalue < 0.05:
    print('p-value가 0.05보다 작으므로 귀무가설을 기각합니다.')
    print('보수 전후 평균 예상 판매가격에는 통계적으로 유의한 차이가 있습니다.')
else:
    print('p-value가 0.05 이상이므로 귀무가설을 기각하지 못합니다.')
    print('보수 전후 평균 예상 판매가격에는 통계적으로 유의한 차이가 있다고 보기 어렵습니다.')


두 배열의 길이 동일 여부 : True

[평균]
보수 전 평균 : 168.5
보수 후 평균 : 177.0

[검정 선택 방법]
대응표본 t검정 진행

[차이값 정규성 검증]
차이값 : [ 9  8  7  9 11  7  8  9  9  8]
검정통계량 : 0.8871 | p-value : 0.1574
판단 : 정규성 가정을 만족합니다.

[대응표본 t검정]
t통계량 : 22.8079
p-value : 0.0000
p-value가 0.05보다 작으므로 귀무가설을 기각합니다.
보수 전후 평균 예상 판매가격에는 통계적으로 유의한 차이가 있습니다.


### 과제 답변 작성란

- **Q1.** 이 데이터는 독립표본인가요, 대응표본인가요?  

    - 대응표본이다.

    - 같은 주택 10채를 보수 전후로 비교하였기 떄문이다.

<br>

- **Q2.** 대응표본 t검정에서는 무엇의 정규성을 확인해야 하나요?  

    - 보수 후 가격과 보수 전 가격의 차이값(after_price - before_price)의 정규성을 확인해야 한다.

<br>

- **Q3.** `stats.ttest_ind()`가 아니라 어떤 함수를 사용해야 하나요?  

    - 대응표본 t검정인 stats.ttest_rel()을 사용해야 한다.

<br>

- **Q4.** 검정 결과 보수 전후 예상 판매가격에 유의한 차이가 있나요?

    - p-value가 0.05보다 작으므로 귀무가설을 기각한다.
    
    - 따라서 보수 전후 평균 예상 판매가격에는 통계적으로 유의한 차이가 있다.

---

## 실습 마무리

1. t통계량은 어떤 두 값을 비교하여 계산하나요?
    - 관측된 평균 차이를 표준오차(그 차이의 불확실성)와 비교한다.

2. 단일표본 t검정은 어떤 상황에서 사용하나요?
    - 하나의 표본 평균을 특정 기준값 또는 모집단 평균과 비교할 때 사용한다.

3. 독립표본과 대응표본은 데이터 수집 구조에서 어떤 차이가 있나요?
    - 독립표본은 서로 다른 대상들로 구성되고, 대응표본은 같은 대상의 반복 측정값처럼 각 관측값이 짝을 이룬다.

4. 독립표본 t검정에서 정규성·등분산성·독립성은 각각 어떻게 확인하나요?
    - 정규성은 Shapiro, 등분산성은 Levene, 독립성은 데이터 수집 구조로 파악한다.

5. 분산이 같다고 보기 어렵거나 확신할 수 없을 때 어떤 t검정을 사용할 수 있나요?
    - equal=True인 Welch t 검정을 사용한다.